libevmasm/Assembly.cpp

```diff
@@ -297,6 +508,132 @@ Json::Value Assembly::assemblyJSON(std::map<std::string, unsigned> const& _sourc
+		solRequire(validMembers.count(attribute), AssemblyImportException, "Unknown attribute '" + attribute + "'.");
+	solRequire(_json.isMember(".code"), AssemblyImportException, "Member '.code' does not exist.");
+	solRequire(_json[".code"].isArray(), AssemblyImportException, "Member '.code' is not an array.");
+	for (auto const& codeItem: _json[".code"])
+		solRequire(codeItem.isObject(), AssemblyImportException, "Item of '.code' array is not an object.");
+
+	if (_level == 0)
+	{
+		if (_json.isMember("sourceList"))
+		{
+			solRequire(_json["sourceList"].isArray(), AssemblyImportException, "Optional member 'sourceList' is not an array.");
+			for (auto const& sourceListItem: _json["sourceList"])
+				solRequire(sourceListItem.isString(), AssemblyImportException, "Item of 'sourceList' array is not of type string.");
+		}
+	} else
+		solRequire(
+			!_json.isMember("sourceList"),
+			AssemblyImportException,
+			"Member 'sourceList' is only allowed in root JSON object."
+		);
+
+	std::vector<std::string> sourceList;
+	if (_json.isMember("sourceList"))
+	{
+		solAssert(_level == 0);
+		for (auto const& it: _json["sourceList"])
+		{
+			solRequire(
+				std::find(sourceList.begin(), sourceList.end(), it.asString()) == sourceList.end(),
+				AssemblyImportException,
+				"Items in 'sourceList' array are not unique."
+			);
+			sourceList.emplace_back(it.asString());
+		}
+	} else
+		sourceList = _sourceList;
+
+	result->importAssemblyItemsFromJSON(_json[".code"], sourceList);
+	if (_json[".auxdata"])
+	{
+		solRequire(_json[".auxdata"].isString(), AssemblyImportException, "Optional member '.auxdata' is not of type string.");
+		bytes auxdata{fromHex(_json[".auxdata"].asString())};
+		solRequire(!auxdata.empty(), AssemblyImportException, "Optional member '.auxdata' is not a valid hexadecimal string.");
+		result->m_auxiliaryData = auxdata;
+	}
+
+	if (_json.isMember(".data"))
+	{
+		solRequire(_json[".data"].isObject(), AssemblyImportException, "Optional member '.data' is not an object.");
+		Json::Value const& data = _json[".data"];
+		for (Json::ValueConstIterator dataIter = data.begin(); dataIter != data.end(); dataIter++)
+		{
+			solRequire(dataIter.key().isString(), AssemblyImportException, "Key inside '.data' is not of type string.");
+			std::string dataItemID = dataIter.key().asString();
+			Json::Value const& code = data[dataItemID];
+			if (code.isString())
+			{
+				if (!code.asString().empty())
+				{
+					bytes data_value{fromHex(code.asString())};
+					solRequire(
+						!data_value.empty(),
+						AssemblyImportException,
+						"Member '.data' contains a value for '" + dataItemID + "' that is not a valid hexadecimal string.");
+				}
+				result->m_data[h256(fromHex(dataItemID))] = fromHex(code.asString());
+			}
+			else if (code.isObject())
+			{
+				size_t index = static_cast<size_t>(std::stoi(dataItemID));
+				if (result->m_subs.size() <= index)
+					result->m_subs.resize(index + 1);
```

Comment:
This looks like it should handle weird indexes, but the import segfaults for me if I try to use a value out of range:
```json
{
    ".code": [],
    ".data": {
        "1": {".code": []}
    }
}
```
```
Segmentation fault (core dumped)
```

Looks like we're missing an assert in some other place that relies on these indexes being in range.

Reply:
True, for this input, the `result->m_subs` will be resized to 2 but the index 0 will have a null pointer, so when the code reaches `result->updatePaths()` it crashes.

Reply:
Does the `result->m_subs` need to have the indexes as the item ID? It seems unnecessary to me, and we could fix this easily by just appending it to the `m_subs` vector instead of inserting it to a specific index. I was not sure of that when I took this PR over, so I kept the logic of what was done initially.

Reply:
The problem is really that we have to stick to the format used by `--asm-json` and that apparently has numeric indexes as IDs. Actually it's even a bit more complex - the keys are numbers if the content is code but can be arbitrary strings if the content is hex data.

I don't think we can do anything about it. I mean, we can always handle more, but we should not change the format in a way that the output from `--asm-json` will not be accepted.

And if the indexes are there, we must use them to order the subs according to them.

Reply:
> The problem is really that we have to stick to the format used by `--asm-json` and that apparently has numeric indexes as IDs. Actually it's even a bit more complex - the keys are numbers if the content is code but can be arbitrary strings if the content is hex data.

Oh, ok, but in the case that the key on `.data` is an arbitrary string, it would not have any subassembly, would it?

> I don't think we can do anything about it. I mean, we can always handle more, but we should not change the format in a way that the output from `--asm-json` will not be accepted.
> 
> And if the indexes are there, we must use them to order the subs according to them.

Hum...so I guess the alternative would be to have `m_subs` as an ordered map instead of a vector.

Reply:
> Oh, ok, but in the case that the key on `.data` is an arbitrary string, it would not have any subassembly, would it?

It wouldn't. Just data. See `if (_json.isMember(".data"))` in `Assembly::fromJSON`.

But looking at it myself now, I wonder if it shouldn't be more strict. Here's the bit that creates `.data` currently and looks like it's not an arbitrary string. Seems to be a hash (of the data name?): https://github.com/ethereum/solidity/blob/b187d065511a5c35d5a9c094bbc2935186b02ee2/libevmasm/Assembly.cpp#L280-L291

I'd adjust the code to accept only the things we can actually produce. We can always relax it later if we introduce something new. A small test covering the validation would be nice too.

> Hum...so I guess the alternative would be to have `m_subs` as an ordered map instead of a vector.

Maybe. I would not be surprised if we were relying on there being no gaps. The simplest thing to do is to just reject any JSON with gaps in numbering as an error and I'd suggest doing just that.

You could start with a local map and only convert it to a vector once you validate it. This would also prevent another issue the current code probably has - that if you use a very large integer for the index, we'll probably try to allocate a very large vector and run out of memory.

Also, please add one case with a gap in numeration (and maybe also a very high index) as a CLI test.

Reply:
> But looking at it myself now, I wonder if it shouldn't be more strict. Here's the bit that creates `.data` currently and looks like it's not an arbitrary string. Seems to be a hash (of the data name?):

Yes, it is the hash of the key.

> I'd adjust the code to accept only the things we can actually produce. We can always relax it later if we introduce something new. A small test covering the validation would be nice too.

Will do.

> Maybe. I would not be surprised if we were relying on there being no gaps. The simplest thing to do is to just reject any JSON with gaps in numbering as an error and I'd suggest doing just that.

But with the addition of the `--import-asm-json` such constraint will not be valid anymore, since you would be able to import a JSON assembly that does not contain all subassemblies, right? Like the one in your example, which starts with ID 1: https://github.com/ethereum/solidity/pull/13673#discussion_r1367256424
 
> You could start with a local map and only convert it to a vector once you validate it. This would also prevent another issue the current code probably has - that if you use a very large integer for the index, we'll probably try to allocate a very large vector and run out of memory.
> 
> Also, please add one case with a gap in numeration (and maybe also a very high index) as a CLI test.

Ok.

Reply:
> But with the addition of the `--import-asm-json` such constraint will not be valid anymore, since you would be able to import a JSON assembly that does not contain all subassemblies, right?

Only if we decide to allow it. What I'm saying is that we don't have to. In fact, I don't see a real use case for this. The numbers are there purely for ordering and you cannot really express those gaps in the bytecode. The produced bytecode should be the same for `"2"`, `"0"`, `"1"` and for `"999"`, `"5"`, `"30"`. It would only make sense to allow the latter if the compiler could produce that numbering. But I'm pretty sure it can't so allowing it just forces us to make the code for handling it more complex without any benefits.